File: src/0090.subsets-ii.90/solution.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */

// 参考 http://www.cnblogs.com/mozi-song/p/9579418.html

var subsetsWithDup = function (nums) {
    // 利用排序把重复元素放在一起
    nums.sort((a, b) => a - b);

    function backTracking (list, nums, sequence, index) {
        sequence.push(list.slice());
        for (let i = index; i < nums.length; i++) {
            // 处理重复元素
            if (i > index && nums[i] === nums[i - 1]) {
                continue;
            }
            list.push(nums[i]);
            backTracking(list, nums, sequence, i + 1);
            list.pop();
        }
    }
    const sequence = [];
    backTracking([], nums, sequence, 0);
    return sequence;
};

File: src/0090.subsets-ii.90/solution.java
class Solution {
    private List<List<Integer>> result = new ArrayList<>();
    boolean[] used;
    private List<Integer> acc = new ArrayList<>();
    int[] nums;
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        used = new boolean[nums.length];
        this.nums = nums;
        backTracking(0);
        return result;
    }

    private void backTracking(int index){
        if(index == used.length){
            result.add(new ArrayList<>(acc));
            return;
        }

        backTracking(index+1);
        
        if(index>0 && nums[index-1] == nums[index]  && !used[index-1]){
            return;
        }


        acc.add(nums[index]);
        used[index] = true;
        backTracking(index+1);
        used[index] = false;
        acc.removeLast();
        
    }

}